src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -0,0 +1,140 @@
+
+public class FirestoreDB implements DBInterface {
+    public String getRecipeContent(String Id) {
+        return null;
+    }
+
+    public RecipeMetadata getRecipeMetadata(String Id) {
+        return null;
+    }
+    public String addRecipe(RecipeMetadata newRecipe, String newContent) {
+        return null;
+    }
+    public void deleteRecipe(String Id) {
+        return;
+    }
+
+    public void editRecipeTitleContent(String editedTitle, String editedContent) {
+        return;
+    }
+
+    public Iterable<RecipeMetadata> getAllRecipes() {
+        return null;
+    }
+    public Iterable<Comment> getAllCommentsInRecipe(String recipeId) {
+        return null;
+    }
+    public Iterable<Tag> getAllTags() {
+        return null;
+    }
+
+    /** Returns a User object from userId. */
+    public User getUser(String userId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      DocumentSnapshot user = DBUtils.blockOnFuture(userRef.get());
+      return user.toObject(User.class);
+    }
+   
+    /** Adds a User to the db based on an input userId. */
+    public String addUser(String userId) {
+      DocumentReference newUserRef = DBUtils.users().document(userId);
+      User newUser = new User(userId);
+      DBUtils.blockOnFuture(newUserRef.set(newUser));  // .set() returns an ApiFuture
+      return userId;
+    }
+    
+    /** Delete a User from the db based on an input userId. */
+    public void deleteUser(String userId) {
+      DBUtils.blockOnFuture(DBUtils.user(userId).delete());
+    }
+    
+    /** Add recipe id: true as a singleton map to user's created recipes field. */
+    public void addRecipeIdToCreated(String userId, String recipeId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      String nestedPropertyName = DBUtils.getNestedPropertyName(User.CREATED_RECIPES_KEY, recipeId);
+      ApiFuture addToCreatedPostsFuture =
+          userRef.update(Collections.singletonMap(nestedPropertyName, true));
+      DBUtils.blockOnFuture(addToCreatedPostsFuture);
+    }
+    
+    /** Remove a recipe from a user's created (when recipe is deleted). */
+    public void removeRecipeIdFromCreated(String userId, String recipeId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      String nestedPropertyName = DBUtils.getNestedPropertyName(User.CREATED_RECIPES_KEY, recipeId);
+      Map<String, Object> update = new HashMap<>();
+      update.put(nestedPropertyName, FieldValue.delete());
+      DBUtils.blockOnFuture(userRef.update(update));
+    }
+    
+    /** Add recipe id: true as a singleton map to user's saved recipes field. */
+    public void addRecipeIdToSaved(String userId, String recipeId) {
+      DocumentReference userRef = DBUtils.user(userId);
+      String nestedPropertyName = DBUtils.getNestedPropertyName(User.SAVED_RECIPES_KEY, recipeId);
+      ApiFuture addToSavedPostsFuture =
+          userRef.update(Collections.singletonMap(nestedPropertyName, true));
```

Comment:
Again, no need for the singleton map here. According the firestore docs, `userRef.update(nestedPropertyName , true)` is sufficient.

Reply:
Okay, good to know, thanks!